### ISSUE TITLE
zune-jpeg: support resizing during decoding?

### ISSUE THREAD

**Issue by Shnatsel** [etemesi254/zune-image#103]
libjpeg-turbo allows getting a lower-resolution version of an image via subsampling during decoding, and jpeg_decoder supports [efficient resizing](https://docs.rs/jpeg-decoder/0.3.0/jpeg_decoder/struct.Decoder.html#method.scale) during decoding. This is useful for creating small thumbnails of large JPEG images much more efficiently.

It would be nice to support something like this in `zune-jpeg`, otherwise integration into `image` would be a regression in memory use compared to `jpeg_decoder`.

Or if you can explain why this is actually a bad idea, that works too :smile:

**Comment by etemesi254** [comment]
I have no problem with implementing it, it makes sense and a lot of image libraaries that use libjpeg-turbo love advertising it(vips, magick)

It is complicated though, as we'd need IDCT for each scaling supported(libjpeg has all 16 of them at https://github.com/libjpeg-turbo/libjpeg-turbo/blob/main/jidctint.c)
and obviously the additional code to support it.

**Comment by Shnatsel** [comment]
Hmm, that is indeed a lot of work. And while the feature is advertised by the libraries, I didn't see it used a whole lot in actual programs, outside certain specific thumbnailers.

Would the subsampling IDCTs perform better than the full-size one, or are the wins purely in memory usage? If it's only memory, then we'd have to also implement the different IDCTs in platform-specific SIMD, which I'm not thrilled about.

**Comment by etemesi254** [comment]
> Would the subsampling IDCTs perform better than the full-size one, or are the wins purely in memory usage? If it's only memory, then we'd have to also implement the different IDCTs in platform-specific SIMD, which I'm not thrilled about.

I don't think so, they would be slower as they aren't vectorized, and the compiler rarely generates vectorized code for such, e.g here is  jpeg-decoders, 4x4 idct https://godbolt.org/z/4TqrP67bE

Ideally, because they use odd sized arrays, 7,5,13 it makes it quite harder for vectorization.

And I've also rarely seen it's usage outside of thumbnails

**Comment by Shnatsel** [comment]
I see. Yes, even thumbnailers often don't use this and just perform a full read and resize. I'm not sure it's worth the trouble then - having an unsafe SIMD implementation for every IDCT size sounds like a lot of unsafe code to write and maintain and verify, and quite a lot of attack surface.

I think this can be reasonably supported once `std::simd` is stabilized, but not before that.

**Comment by sophie-h** [comment]
Loupe, which hopefully will become GNOME's new default image viewer soon, uses `image-rs` for most formats. Ideally, we would have support for both, getting a downscaled version of JPEGs and choosing just an area of the image.

The background is that larger images might hit RAM and, more likely, VRAM limits (we use GL for rendering when available). Therefore downscaling would be required to show huge images (huge relative to the hardware.) Choosing just an area of the image could be used for feeding the result into our tiled renderer, also allowing us to zoom into large images with limited resources.

So we would indeed use those features.

**Comment by Shnatsel** [comment]
I'm glad to hear that image-rs is making its way into more places! ❤️ 

Loupe would still need a generic solution for not running out of VRAM for all the other formats where this kind of optimization isn't possible, such as PNG. So while it would certainly be nice to have this one special-cased format to output images that are already somewhat downscaled, it won't solve the general problem.

**Comment by Shnatsel** [comment]
Also, I understand scaling in the JPEG decoder would actually be slower than decoding the full-size image up front in case the user ever zooms in, because the entire image would have to be decoded again into its full size, meaning it's decoded twice.

I understand the GPU can scale images very efficiently (mipmapping is basically that), so wouldn't the fastest option be slicing the image into tiles, sending a batch of them to the GPU, downscaling them on the GPU and then sending the next batch until the whole image is downscaled? For extreme cases where `last_tile_batch + downscaled image` don't fit into VRAM this process can fall back to software downscaling at the end.

**Comment by sophie-h** [comment]
Yeah, maybe this is unnecessary optimization.

**Comment by kornelski** [comment]
Maybe as a compromise you could support DC-only decoding? 1/8th resolution. That's still immensely useful for thumbnails, and just skips DCT entirely.

**Comment by etemesi254** [comment]
> Maybe as a compromise you could support DC-only decoding?
Yes, I can skip DCT if requested by a flag by propagating the DC value

> 1/8th resolution.
Too complicated.

It's another path, in an already complicated path, it will invalidate some logic that will definitely make something else crash and probably won't be well fuzz tested, will cause crashes, and headaches debugging why suddenly we have a zero width image (because we did `width/8` and not `(width+7)/8` )

While I get why you'd wan't to skip DCT,i.e decoding becomes faster, i'd rather you use traditional resizing algorithms for resizing, I very much would like `zune-jpeg` to remain as a jpeg-decoder and not jpeg-decoder + resizer.

**Comment by sophie-h** [comment]
We are planning to switch the complete image stack in GNOME over to Rust as far as possible.

I am now working on a thumbnailer and using zune-image instead of libjpeg gives me a performance hit in the range of 2x to 5x due to zune-image not being capable of downscaling during loading.

This is somewhat frustrating since for viewing images in an app, using zune-jpeg is considerably faster.

Could this feature maybe be reconsidered? Thumbnailing is quite an important feature for things like gallery apps or websites etc.

**Comment by Shnatsel** [comment]
Resizing during decoding is not any faster, at least on the CPU; it just uses less memory. There are some savings to be had by not loading the decoded image in memory, but a 5x hit seems excessive. Could you record a profile of the thumbnailing operation with [samply](https://crates.io/crates/samply), click "upload profile" in the top right corner and share the link here?

What I suspect is taking so much time is the expensive downscaling afterwards, not the decoding itself. You can combat this by running a very cheap subsampling to 4x the target resolution, then run the proper expensive downscaling for the last leg of it. This is what `imagemagick -thumbnail` does, and [my replication of it in wondermagick](https://github.com/Shnatsel/wondermagick/blob/7e0c3eedfefe59839f8113c87718aa82b3224cee/src/operations/resize.rs#L34-L50) is very fast, outperforming imagemagick itself by 1.6x.

**Comment by Shnatsel** [comment]
Hmm, I did more experiments with it and I think I was wrong. 

I tried a more advanced imagemagick invocation that uses the resize-while-decoding feature of libjpeg-turbo and it does beat wondermagick handily in that configuration:

```
hyperfine --warmup=3 'wm-convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg' 'convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg' 'convert -define jpeg:size=240x240 ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg'
Benchmark 1: wm-convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg
  Time (mean ± σ):     103.4 ms ±   2.4 ms    [User: 78.2 ms, System: 24.9 ms]
  Range (min … max):    98.0 ms … 108.1 ms    27 runs
 
Benchmark 2: convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg
  Time (mean ± σ):     162.1 ms ±   3.5 ms    [User: 113.8 ms, System: 56.8 ms]
  Range (min … max):   157.3 ms … 167.6 ms    18 runs
 
Benchmark 3: convert -define jpeg:size=240x240 ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg
  Time (mean ± σ):      60.2 ms ±   1.1 ms    [User: 61.2 ms, System: 4.9 ms]
  Range (min … max):    58.9 ms …  62.6 ms    48 runs
 
Summary
  convert -define jpeg:size=240x240 ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg ran
    1.72 ± 0.05 times faster than wm-convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg
    2.69 ± 0.08 times faster than convert ~/Sun_over_Lake_Hawea,_New_Zealand.jpg -thumbnail 120x120 ~/out.jpg
```

It seems like some (limited) support for resizing during decoding may be necessary after all. Perhaps the approach of skipping DCT, like @kornelski suggested?

**Comment by sophie-h** [comment]
Yes, the speed of libjpeg-turbo heavily depends on the targeted thumbnail size. Which is no surprise with its mentioned support for IDCT for each sampling option.

Of course, I don't need zune-jpeg to take care of the scaling to the exact size, just to reduce the amount of work needed.

> What I suspect is taking so much time is the expensive downscaling afterwards

It's taking considerable time, but no way to match libjpeg-turbo with just optimizations there.

**Comment by sophie-h** [comment]
> is there some official-ish place where this intent is stated

Yes, I already mentioned it in the image-rs Matrix channel: https://gitlab.gnome.org/GNOME/Initiatives/-/issues/53

Feel free to quote this for funding etc. If you need more, let me know. Hopefully I will write a blog post in the next weeks as well.

**Comment by kornelski** [comment]
libjpeg only supports n/8 scaling factors, which maps to DCT blocks of a given size (like 3x3 blocks instead of 8x8). The image still needs another resizing step when an exact size is needed. 

For my usage it's always faster — fewer pixels decoded, fewer pixels to resize afterwards. I have to note that I'm using DCT scaling only for resizing down to >=1.5x of the final size in order to apply final anti-aliasing using a better filter.

The smaller the output the more difference it makes, so 7/8 scaling is not very useful, but 4/8 and 2/8 are nice, and as I've mentioned previously, 1/8 is the easiest and the most beneficial case.

**Comment by Shnatsel** [comment]
I've just checked how the 1/4 scaling performs in imagemagick and turns out wondermagick still beats it even without supporting scaling during JPEG decoding. So it does seem like supporting the 1/8 scaling is all we really need to enable thumbnailing use cases.

**Comment by kornelski** [comment]
The point is in doing less IDCT work. That's a non-trivial part of the decoding process, so it's silly to IDCT 64 pixels (with chroma upscaling and YUV conversion) only to use 1 or 4 of them.

**Comment by dzamlo** [comment]
For anybody needing this, it's possible to use the `libjpeg-turbo` via the `turbojpeg` crate if it works for your use case: https://docs.rs/turbojpeg/1.3.3/turbojpeg/struct.Decompressor.html#method.set_scaling_factor
